### PR TITLE
WIP: Add support for autoencoding variational Bayes

### PR DESCRIPTION
**Pull request by taku-y**
Autoencoding VB (AEVB) has been implemented for mini-batch ADVI. An example notebook for latent dirichlet allocation is also added. `advi_minibatch()` has been moved to a separate file. Motivation of AEVB is to deal with local latent variables, each of which is associated with a given sample. When the number of samples is large, encoding variational parameters from given samples is computationally efficient.

### CODE REVIEW

File: pymc3/tests/test_advi.py [hunk 1]
```diff
@@ -2,7 +2,8 @@
 import pymc3 as pm
 from pymc3 import Model, Normal, DiscreteUniform, Poisson, switch, Exponential
 from pymc3.theanof import inputvars
-from pymc3.variational.advi import variational_gradient_estimate, advi, advi_minibatch, sample_vp
+from pymc3.variational import advi, advi_minibatch, sample_vp
```
Comment (twiecki) [inline pymc3/tests/test_advi.py:5]: Do we need to fix any other examples for the import?

File: pymc3/variational/advi_minibatch.py [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+    start = {v.name: start[v.name] for v in global_RVs}
+    bij = DictToArrayBijection(global_order, start)
+    u_start = bij.map(start)
+    w_start = np.zeros_like(u_start)
+    uw_start = np.concatenate([u_start, w_start])
+    uw_global_shared = theano.shared(uw_start, 'uw_global_shared')
+
+    return uw_global_shared, bij
+
+# def _join_RVs(global_RVs, global_order, local_RVs, local_order):
```
Comment (twiecki) [inline pymc3/variational/advi_minibatch.py]: Can we delete this?

File: pymc3/variational/advi_minibatch.py [hunk 3]
```diff
@@ -0,0 +1,469 @@
+
+'''
+Created on Mar 12, 2011
+
+@author: johnsalvatier
```
Comment (twiecki) [inline pymc3/variational/advi_minibatch.py]: Fix copyright date and author.

**Comment by twiecki** [comment]
Looks great. Can we remove the commented out code?

Do you think we should replace your `MyDirichlet` with our current one? What are the differences? Same question re StickBreaking transform in your NB.

File: pymc3/tests/test_advi.py (hunk 1 above)
Comment (taku-y) [inline pymc3/tests/test_advi.py:5]: I will check examples importing advi_minibatch() and fix if needed.

File: pymc3/variational/advi_minibatch.py (hunk 2 above)
Comment (taku-y) [inline pymc3/variational/advi_minibatch.py]: This is used in advi_minibatch().

File: pymc3/variational/advi_minibatch.py (hunk 3 above)
Comment (taku-y) [inline pymc3/variational/advi_minibatch.py]: I will fix it.

**Comment by taku-y** [comment]
> > Looks great. Can we remove the commented out code?

I will remove comments. 

> > Do you think we should replace your MyDirichlet with our current one? What are the differences? Same question re StickBreaking transform in your NB.

I think we should change them. 

For `MyDirichlet`, the order of dimensions are changed. Specifically, rows and columns of the random variables flipped. This change probably breaks other examples where Dirichlet is used. I think we should fix this in #1125. 

For both of `MyDirichlet` and `TruncatedStickBreaking`, random variables are truncated so as not to take value 0 for avoiding numerical issues like #1147 (and fixed in #1156). The truncation can be optional, which is used like `Dirichlet(offset=1e-6)`, because it is a workaround for computation and users should be aware it.

**Comment by twiecki** [comment]
> For MyDirichlet, the order of dimensions are changed. Specifically, rows and columns of the random variables flipped. This change probably breaks other examples where Dirichlet is used. I think we should fix this in #1125.

It's unclear when #1125 will be ready so I think we should fix the Dirichlet earlier than that. I can do a PR for that.

> For both of MyDirichlet and TruncatedStickBreaking, random variables are truncated so as not to take value 0 for avoiding numerical issues like #1147 (and fixed in #1156). The truncation can be optional, which is used like Dirichlet(offset=1e-6), because it is a workaround for computation and users should be aware it.

I see. Well I think we should definitely replace that as well then, the more numerical robustness of these methods the better. This could be part of this PR?

**Comment by twiecki** [comment]
I ported your fixes over to https://github.com/pymc-devs/pymc3/pull/1241

File: pymc3/tests/test_advi.py (hunk 1 above)
Comment (taku-y) [inline pymc3/tests/test_advi.py:5]: I checked that the below notebooks work with small fixes. 
- docs/source/notebooks/bayesian_neural_network_advi.ipynb
- docs/source/notebooks/gaussian-mixture-model-advi.ipynb
- docs/source/notebooks/GLM-hierarchical-advi-minibatch.ipynb

Comment (twiecki) [inline pymc3/tests/test_advi.py:5]: :+1:

**Comment by taku-y** [comment]
> I see. Well I think we should definitely replace that as well then, the more numerical robustness of these methods the better. This could be part of this PR?

I think this should be a different PR. I will do after #1241 is merged.

**Comment by twiecki** [comment]
> I think this should be a different PR. I will do after #1241 is merged.

I included the stick-breaking there as well now. Is it truncated? I can only see the axes change.

**Comment by taku-y** [comment]
In the notebook, I redefined `inverse_logit()`

```
upper = 1 - 1e-9
lower = 1e-9
inverse_logit = lambda x: (upper - lower) * tt.nnet.sigmoid(x) + lower
```

**Comment by twiecki** [comment]
How is that better / different from:

``` python
def invlogit(x):
    p_min = sys.float_info.epsilon

    return (1 - 2 * p_min) / (1 + tt.exp(-x)) + p_min
```

**Comment by taku-y** [comment]
It looks good to me.

**Comment by twiecki** [comment]
That's what we have right now in `math.py`. Just wondering why you changed it here.

**Comment by taku-y** [comment]
Sorry for the confusion. I wanted to make upper and lower bounds clear in the new code, but either way is OK.

**Comment by twiecki** [comment]
Now that #1241 is merged the code here could be simplified a bit.

**Comment by taku-y** [comment]
I rebased the code and pushed with '-f' option as described in https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request, but Github say "This branch has conflicts that must be resolved". I will solve the problem tomorrow.

**Comment by twiecki** [comment]
This is awesome. I wonder if in the LDA the ADVI has converged as it's still going up? Anyway, I'll go ahead and merge this but maybe something to follow up on.

**Comment by taku-y** [comment]
I didn't noticed that the result of ADVI was poor than before. I fixed the learning rate and increased iterations for better result, see https://gist.github.com/taku-y/66c9613ab29a150e4493b899a6507354. The ELBO looks converged.

**Pull request merged**